synthesizer/src/vm/tests/test_v10.rs

```diff
@@ -0,0 +1,411 @@
+// Copyright (c) 2019-2025 Provable Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+use crate::vm::test_helpers::*;
+
+use console::{account::ViewKey, program::Value};
+use snarkvm_synthesizer_program::Program;
+
+use console::network::ConsensusVersion;
+use snarkvm_utilities::TestRng;
+
+// This test verifies that:
+//  - a dependency cycle can be created between two programs that use records in their calls.
+//  - a program with a call cycle cannot be deployed.
+//  - a program that changes an output record index cannot be deployed.
+//  - a program that returns a local record as an external record cannot be deployed.
+//  - a record cannot be consumed twice in the same program.
+//  - a local record cannot be created and consumed in the same program.
+#[test]
+fn test_upgrade_cycle_with_records() {
+    let rng = &mut TestRng::default();
+
+    // Initialize a new caller.
+    let caller_private_key = crate::vm::test_helpers::sample_genesis_private_key(rng);
+    let caller_view_key = ViewKey::try_from(&caller_private_key).unwrap();
+
+    // Initialize the VM at the V10 height.
+    let v10_height = CurrentNetwork::CONSENSUS_HEIGHT(ConsensusVersion::V10).unwrap();
+    let vm = crate::vm::test_helpers::sample_vm_at_height(v10_height, rng);
+
+    // Define the first program with a record.
+    let program_one = Program::from_str(
+        r"
+program test_one.aleo;
+
+record First:
+    owner as address.private;
+    data as field.private;
+
+constructor:
+    assert.eq true true;
+
+function mint:
+    input r0 as field.private;
+    cast self.caller r0 into r1 as First.record;
+    output r1 as First.record;
+
+function first:
+    input r0 as First.record;
+    cast r0.owner r0.data into r1 as First.record;
+    output r1 as First.record;
+",
+    )
+    .unwrap();
+
+    // Define the second program with a record that calls the first program.
+    let program_two = Program::from_str(
+        r"
+import test_one.aleo;
+
+program test_two.aleo;
+
+record Second:
+    owner as address.private;
+    data as field.private;
+
+constructor:
+    assert.eq true true;
+
+function second:
+    input r0 as test_one.aleo/First.record;
+    call test_one.aleo/first r0 into r1;
+    cast r0.owner r0.data into r2 as Second.record;
```

Comment:
By the way, this behaviour is also somewhat surprising to me: apparently we can read previously consumed record values, which kind of messes with the "pass by value" analogy. Do you think this is desired? I'm happy to open a new issue on snarkVM or our docs based on our assessment.